### PR TITLE
api: MetaProtocol Proxy

### PR DESCRIPTION
**Pull request by zhaohuabing**
This PR proposes a MetaProtocol proxy that provides a framework for layer-7 protocols. The common capabilities, including routing, tracing, metrics, logging, etc., will be built into the MetaProtocol proxy. This approach has the potential to significantly lower the barrier to writing new Envoy filters: instead of writing fully functional network filters and duplicating similar codes, now we only need to implement the codec interface. Furthermore, MetaProtocol makes it much easier to build a unified control plane and it could be very helpful for the service mesh ecosystem.

Screen Shot 2021-10-29 at 4 09 49 PM

This PR is a merged efforts of https://github.com/envoyproxy/envoy/pull/18768 and https://github.com/envoyproxy/envoy/issues/18209 

This PR is the APIs of the MetaProtocol Proxy. The implementation will be pushed later after the community agrees on the design and finalizes the API. 

Related issues: 
Common framework for layer-7 protocols #18761
generic proxy support: #18209
Generic Access Log: #13085
Generic Routing & Rds: #9656

Commit Message: Add API for MetaProtocol Proxy
Additional Description:
Design doc: https://docs.google.com/document/d/1y_ALyjGp_H1_TRrrRxZxnqENISqeZI6eLn71FAFsZPk/edit?usp=sharing
Istio blog about MetaProtocol: https://istio.io/latest/blog/2021/aeraki/
This design has been proven working in Aeraki : https://github.com/aeraki-framework/meta-protocol-proxy
Risk Level: low
Testing:
Docs Changes:  
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by htuch** [comment]
Thanks @zhaohuabing @wbpcode. As mentioned before, I'm highly supportive of this work and this seems a solid architectural pattern. I've left a first round of comments on your doc; let's make sure we're all aligned on the details there before diving into code review.

**Comment by wbpcode** [comment]
@htuch Thanks very much. 🌷 We will update our documentation as soon as possible.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:33]: What's the long term plan here? How much of the existing `RouteAction` in RDS will be reproduced? If it's the entire thing, then arguably we should just augment existing `RouteAction` with the ability to match on MetaProtocol attributes (maybe it can already via dynamic metadata matchers?). If not, we should be clear on the long term relationship across these structures.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:16]: How are these matchers used? It would be good to see some examples of config fragments somewhere.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:36]: This is the same discussion we had in https://github.com/envoyproxy/envoy/pull/17631 which led to TRDS; if we are reusing `Rds` here, we should be shipping the `RouteConfiguration` as it exists today. Otherwise we need a MetadataRDS. CC @markdroth

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:80]: This should all be `TypedExtensionConfig`

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:70]: `TypedExtensionConfig`

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/route.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/route.proto:25]: Nice to see the switch to this (CC @snowp). Have you a working implementation yet based on these matchers? Any reference PR with all the pieces connected together?

In general the strategy of small incremental PRs is good, but it's also good to have some branch or PR to refer to that connects this.

Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/route.proto:25]: Thanks. We have implemented part of it and plan to submit a new PR after the API is merged. All progress and PR will be recorded in #18209.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:33]: We will check it.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:16]: Get it.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:80]: get it.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:70]: get it.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:16]: > How are these matchers used? It would be good to see some examples of config fragments somewhere.

@htuch examples can be found in the design doc: https://docs.google.com/document/d/1y_ALyjGp_H1_TRrrRxZxnqENISqeZI6eLn71FAFsZPk/edit#heading=h.pgxyh6f6oib4

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:33]: @htuch @wbpcode 

I had the same thought about that: should we reuse the existing RouteAction? Or create another similar one? 

My concern is the existing RouteAction is highly coupled with HTTP protocols. If we reuse it, some properties or inner structures may become ambiguous for MetaProtocol. 

```
{
  "cluster": "...",
  "cluster_header": "...",
  "weighted_clusters": "{...}",
  "cluster_not_found_response_code": "...",
  "metadata_match": "{...}",
  "prefix_rewrite": "...",
  "regex_rewrite": "{...}",
  "host_rewrite_literal": "...",
  "auto_host_rewrite": "{...}",
  "host_rewrite_header": "...",
  "host_rewrite_path_regex": "{...}",
  "timeout": "{...}",
  "idle_timeout": "{...}",
  "retry_policy": "{...}",
  "request_mirror_policies": [],
  "priority": "...",
  "rate_limits": [],
  "include_vh_rate_limits": "{...}",
  "hash_policy": [],
  "cors": "{...}",
  "max_grpc_timeout": "{...}",
  "grpc_timeout_offset": "{...}",
  "upgrade_configs": [],
  "internal_redirect_policy": "{...}",
  "internal_redirect_action": "...",
  "max_internal_redirects": "{...}",
  "hedge_policy": "{...}",
  "max_stream_duration": "{...}"
}
```

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto]: @htuch per what we disscussed in the design docs, it's been renamed as MetaRds.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:81]: fixed

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto]: fixed

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:33]: @zhaohuabing I think so too. May be we should create new RouteAction which is protocol-independent for the meta protocol proxy.  cc @htuch

**Comment by rojkov** [comment]
To fix formatting automatically you can run `./tools/code_format/check_format.py fix`. Also it's possible to install git hooks doing prechecks locally upon git push. The hooks can be installed with ` ./support/bootstrap`.

**Comment by zhaohuabing** [comment]
> To fix formatting automatically you can run `./tools/code_format/check_format.py fix`. Also it's possible to install git hooks doing prechecks locally upon git push. The hooks can be installed with ` ./support/bootstrap`.

Thanks for the hint @rojkov

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:34]: I agree it's not great to have the duplication, but I would be OK starting from scratch and making sure we don't have ambiguity.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (mattklein123) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:18]: Would it be possible in this PR to have some config snippets? somewhere? I think that would really help.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/README.md
Comment (mattklein123) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/README.md:1]: nit: delete, we don't need this.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:38]: You could potentially put examples here, or in the proto docs for RouteConfiguration.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this is super cool. Overall this LGTM but per some comments it would be nice to have some more fully formed config examples we can look at?

/wait

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:18]: An example of the meta protocol route configuration with matcher API has been added to the route.proto.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:38]: An example of the meta protocol route configuration with matcher API has been added to the route.proto.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/README.md
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/README.md]: Deleted.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto:45]: remove `aeraki.meta_protocol.filters.router`

**Review by wbpcode — CHANGES_REQUESTED** [review]
some nit comments.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto]: Removed.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/v3/meta_protocol_proxy.proto]: Fixed.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto
Comment (zhxie) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:34]: https://github.com/cncf/xds/blob/cb28da3451f158a947dfc45090fe92b07b243bc1/xds/type/matcher/v3/matcher.proto#L38-L39

I think different protocols differ on `RouteAction`, would it be better if we can provide a public interface similar to `action` in the generic matching?

Comment (wbpcode) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/action/v3/action.proto:34]: If it's necessary, we can add new `custom_protocol_specific_config` field to the `RouteAction` in the future.

**Comment by zhaohuabing** [comment]
Hi, guys,

Is there anything more I can do to move this forward?

**Review by mattklein123 — DISMISSED** [review]
Thanks this LGTM and given that it's marked WIP I think we can ship and iterate as needed. @htuch @snowp any further comments on this one?

**Comment by htuch** [comment]
SGTM to merge as WiP, defer to @snowp for final look.

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (snowp) [inline api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:21]: Are service and method just special properties that we think are common enough to have specific types? Or is there something else that makes them special?

I would probably have the comment be a bit more vague here if they are just special properties, like `Only applicable if a method provided by the application protocol`. gRPC for example is an RPC protocol that doesn't use methods (outside of google.api.annotations)

**Review by snowp — COMMENTED** [review]
Just a small comment, this looks great to me!

File: api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto
Comment (zhaohuabing) [reply api/envoy/extensions/filters/network/meta_protocol_proxy/matcher/v3/matcher.proto:21]: @snowp your concern is legitimate, not every RPC protocol has these two properties.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
